### ISSUE TITLE
Trouble with GraalVM native-image post compile

### ISSUE THREAD

**Issue by EasyG0ing1** [classgraph/classgraph#860]
Hello,

I'm trying to build a project that uses ClassGraph into a GraalVM native-image.

GraalVM does not handle reflection dynamically, so when code uses reflection, those references have to be explicitly stated in a json file so that Graal can make those connections during the compile phase.

GraalVM also offers a handy tracer option that one can use where it monitors reflection use of your code as you run it and put it through its paces etc. where it automatically builds the necessary json files that you then include in your build statement when you compile the code.

The issue I'm running into seems to be with the way that ClassGraph does things so dynamically. My assumption is that there might be some nested dynamic reflection going on that the tracer tool can't catch and I'm not sure how to locate those so that I can add them into the proper json file.

Here is the stack trace that comes out of the native image once it's compiled:

```bash
Exception in Application start method
Exception in thread "main" java.lang.RuntimeException: Exception in Application start method
        at javafx.graphics@22.0.1/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:893)
        at javafx.graphics@22.0.1/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:196)
        at java.base@22.0.1/java.lang.Thread.runWith(Thread.java:1583)
        at java.base@22.0.1/java.lang.Thread.run(Thread.java:1570)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:858)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:834)
Caused by: java.lang.ExceptionInInitializerError
        at SimpleOTP@1.0/com.simtechdata.simpleotp.database.DB.<clinit>(DB.java:27)
        at SimpleOTP@1.0/com.simtechdata.simpleotp.Main.tryDatabase(Main.java:153)
        at SimpleOTP@1.0/com.simtechdata.simpleotp.Main.start(Main.java:125)
        at javafx.graphics@22.0.1/com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:839)
        at javafx.graphics@22.0.1/com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:483)
        at javafx.graphics@22.0.1/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:456)
        at java.base@22.0.1/java.security.AccessController.executePrivileged(AccessController.java:128)
        at java.base@22.0.1/java.security.AccessController.doPrivileged(AccessController.java:400)
        at javafx.graphics@22.0.1/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:455)
        at javafx.graphics@22.0.1/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
Caused by: io.github.classgraph.ClassGraphException: Uncaught exception during scan
        at io.github.classgraph/io.github.classgraph.ClassGraph.scan(ClassGraph.java:1637)
        at io.github.classgraph/io.github.classgraph.ClassGraph.scan(ClassGraph.java:1654)
        at io.github.classgraph/io.github.classgraph.ClassGraph.scan(ClassGraph.java:1667)
        at SimpleOTP@1.0/com.simtechdata.simpleotp.resources.Resources.<init>(Resources.java:39)
        at SimpleOTP@1.0/com.simtechdata.simpleotp.resources.Resources.<clinit>(Resources.java:30)
        ... 10 more
Caused by: java.lang.IllegalArgumentException: Method "list" could not be invoked: java.lang.reflect.InvocationTargetException
        at io.github.classgraph/nonapi.io.github.classgraph.reflection.ReflectionUtils.invokeMethod(ReflectionUtils.java:231)
        at io.github.classgraph/io.github.classgraph.ModuleReaderProxy.list(ModuleReaderProxy.java:110)
        at io.github.classgraph/io.github.classgraph.ClasspathElementModule.scanPaths(ClasspathElementModule.java:304)
        at io.github.classgraph/io.github.classgraph.Scanner$4.processWorkUnit(Scanner.java:1097)
        at io.github.classgraph/io.github.classgraph.Scanner$4.processWorkUnit(Scanner.java:1091)
        at io.github.classgraph/nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkLoop(WorkQueue.java:246)
        at io.github.classgraph/nonapi.io.github.classgraph.concurrency.WorkQueue.access$000(WorkQueue.java:50)
        at io.github.classgraph/nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:201)
        at io.github.classgraph/nonapi.io.github.classgraph.concurrency.WorkQueue$1.call(WorkQueue.java:198)
        at java.base@22.0.1/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base@22.0.1/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base@22.0.1/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base@22.0.1/java.lang.Thread.runWith(Thread.java:1583)
        at java.base@22.0.1/java.lang.Thread.run(Thread.java:1570)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:858)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:834)
```

Based on another issue that I saw here related to GraalVM, I think we are both kind of treading on uncharted waters so to speak. So I'm starting this issue with hopes that together, we can figure out how to properly compile a java project into a native-image when ClassGraph is used in the project.

My project is modular by the way and I'm using Java 22.

I am able to run the project from its compiled jar file without any issue it's only after compiling it into a native-image do I run into this issue.

Thanks.

**Comment by lukehutch** [comment]
Hi, I'm a huge fan of GraalVM's usage of partial evaluation (Futamura projections) for cross-platform compilation, but I have never tried to use it for anything serious, so I really don't know anything about this.

By the looks of things though, the Graal runtime does not compile the Java module system into the final binary (it may just resolve modules and classes at compiletime), and/or has limitations on reflection, as you point out.

The reason ClassGraph uses reflection to call `ModuleReader` (via `ModuleReaderProxy`) is that ClassGraph is built for JDK 7, for maximum compatibility. I have considered raising the minimum JDK version to 11 or even 17 for ClassGraph 5.0 (although I don't have time to work on that right now). But part of that change could be to drop `ModuleReaderProxy` and just use `ModuleReader` (which was introduced in JDK 9) directly.

What I wrote in the docs is that to support Java environments that are non-standard (like Android, or GraalVM), you should do your scanning at build-time:

https://github.com/classgraph/classgraph/wiki/Build-Time-Scanning

**Comment by EasyG0ing1** [comment]
@lukehutch 

I tried serializing everything at build time then added the json file into my resources folder, then built the ScanResult from the json file, but I get the same results.

I'm using ClassGraph mainly for the convenience of getting at the files in the resources folder. It's been handy for this project where those files change a lot, ClassGraph lets me not have to worry about tracking the files explicitly in the program.

I may need to just go back to Javas built in means of getting at those files. Graal seems to handle that method pretty well.

**Comment by lukehutch** [comment]
OK, well if you can add the resource files to the traditional classpath, rather than to the module path, then you can disable module scanning, and ClassGraph should work for you again.

**Comment by lukehutch** [comment]
We'll be moving to a minimum Java version of >9 for ClassGraph v5, so `ModuleReaderProxy` will be removed and `ModuleReader` will be used directly, which will fix this. Thanks for the report.
